### ISSUE TITLE
Better Twig debugging?

### ISSUE THREAD

**Issue by becw** [palantirnet/the-build#20]
Implement the full setup from https://www.drupal.org/node/1903374 on development, and make sure that it doesn't carry through to prod?

**Comment by froboy** [comment]
Long time, no see, friendo. I implemented this on AMA at https://github.com/AmericanMedicalAssociation/ama-d8/pull/44/commits/d4c4dc98624c72f5dc6953160569a91242920c8e

For those who can't see that repo, here's the diff:

```
From d4c4dc98624c72f5dc6953160569a91242920c8e Mon Sep 17 00:00:00 2001
From: Avi Schwab <schwab@palantir.net>
Date: Thu, 31 Aug 2017 17:27:10 -0500
Subject: [PATCH] Make sure twig debugging is really really on. as per
 https://www.drupal.org/node/2598914

---
 provisioning/templates/services.j2.yml | 11 ++++++++---
 provisioning/templates/settings.j2.php |  5 +++++
 2 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/provisioning/templates/services.j2.yml b/provisioning/templates/services.j2.yml
index 55ebddb..4179b5a 100644
--- a/provisioning/templates/services.j2.yml
+++ b/provisioning/templates/services.j2.yml
@@ -6,9 +6,9 @@ parameters:
     gc_maxlifetime: 200000
     cookie_lifetime: 2000000
   twig.config:
-    debug: {{ item.twig_debug }}
-    auto_reload: null
-    cache: true
+    debug: {{ item.twig_debug|lower }}
+    auto_reload: {{ 'true' if item.twig_debug else 'null' }}
+    cache: {{ (not item.twig_debug)|lower }}
   renderer.config:
     required_cache_contexts: ['languages:language_interface', 'theme', 'user.permissions']
     auto_placeholder_conditions:
@@ -34,3 +34,8 @@ parameters:
     - sftp
     - webcal
     - rtsp
+{% if item.twig_debug %}
+services:
+  cache.backend.null:
+    class: Drupal\Core\Cache\NullBackendFactory
+{% endif %}
diff --git a/provisioning/templates/settings.j2.php b/provisioning/templates/settings.j2.php
index e2a5547..7a5b448 100644
--- a/provisioning/templates/settings.j2.php
+++ b/provisioning/templates/settings.j2.php
@@ -2,6 +2,11 @@
 
 $settings['cache']['default'] = 'cache.backend.database';
 
+{% if item.twig_debug %}
+$settings['cache']['bins']['render'] = 'cache.backend.null';
+$settings['cache']['bins']['dynamic_page_cache'] = 'cache.backend.null';
+
+{% endif %}
 $settings['hash_salt'] = '{{ item.hash_salt }}';
 $settings['update_free_access'] = FALSE;
 $settings['container_yamls'][] = __DIR__ . '/services.yml';
```
we've got some new tools there, but this shouldn't be too hard to integrate back into the build. I'll try and do it some time soon.

**Comment by froboy** [comment]
Note: this is moot for flat repos. Just set those settings in the `services.dev.yml` and `settings.dev.php` or whatever the corresponding files are.

**Comment by froboy** [comment]
Yea, working with this for a while I think it makes a lot of sense for these all to be on by default. I don't think I've ever seen significant performance degradation to a new D8 site with all the dev things on locally. I think some of the instructions for turning it all on might have been simplified a bit since the issue was opened.

For clarity, I think this is what I'm talking about when I say "all the debugging things":
- CSS/JS aggregation is turned off (this is prob a different ticket)
- twig debugging is on
- twig caching is off
- twig auto-reload is on (these three should mean the _only_ reason to need `drush cr` is for new files)
- `http.response.debug_cacheability_headers` is true

**Comment by becw** [comment]
Merged to `release-2.0`!
